<html>
      <p align="center"><font size="+2" face="Arial">primer_match</font></p>

<h3>Name</h3>
<blockquote>
<p><b>primer_match</b> - Find and count primers in a DNA
sequence database</p>
</blockquote>
<h3>Synopsis</h3>
<blockquote>
<p><b>primer_match</b> [
<i> options</i> ]</p>
</blockquote>
<h3>Description</h3>
<blockquote>
<p><b>primer_match</b> finds and counts exact and near exact instances of short
DNA sequences, usually primers, in a (much) larger DNA sequence database such as the human genome.&nbsp;</p>
<p>By default, <b> primer_match</b> outputs a human readable
alignment for each occurance of a primer in the sequence database. With
appropriate option, <u>-c</u>, <b>primer_match</b> will output the number of
occurances of each primer. The format of the alignments and counts is completely configurable with the
<u> -A</u> and <u> -C</u> options.</p>
<p><b>primer_match </b>runs fastest when the sequence database has been
pre-processed with <b>compress_seq</b>, but this is not necessary. If the
sequence database has not been pre-processed with <b>compress_seq</b>, the
sequence database <u>must</u> be in a regular FASTA format. Each line, except
for the last, of every sequence entry must hold the same number of sequence
characters. If the sequence database is not in a regular FASTA format, the
results may be incorrect. <b>primer_match</b> will warn the user if the FASTA
format is not in a regular format.&nbsp;</p>
</blockquote>
<h3>Options</h3>
<blockquote>
<p>-i <i>FASTA</i><i>_sequence_database</i></p>
  <blockquote>
<p>Name of the sequence database to search. Required.</p>
  </blockquote>
<p>-p <i> primers</i></p>
  <blockquote>
<p>White space (space, tab, new line) separated list of primer sequences to find in the sequence database. When this
opition is used on the command line, the primers will usually need to be placed in quotes (&quot;). One of
<u>-p</u>, <u>-P</u>, <u>-F, </u>or <u>-S</u> must be supplied.</p>
  </blockquote>
<p>-P <i>primer_file</i></p>
  <blockquote>
<p>File containing a white space (space, tab, new line) separated list of primer 
sequences to find in the sequence database. One of
<u>-p</u>, <u>-P</u>, <u>-F, </u>or <u>-S</u> must be supplied.</p>
  </blockquote>
<p>-F <i>fasta_primer_file</i></p>
  <blockquote>
<p>FASTA file containing a list of primer sequences to find in the sequence 
database. One of
<u>-p</u>, <u>-P</u>, <u>-F, </u>or <u>-S</u> must be supplied.</p>
  </blockquote>
<p>-S <i>sts-format-file</i></p>
  <blockquote>
<p>UniSTS format file containing a list of primer sequences to find in the 
sequence database. One of
<u>-p</u>, <u>-P</u>, <u>-F, </u>or <u>-S</u> must be supplied.</p>
  </blockquote>
<p>-o <i> output_file</i></p>
  <blockquote>
<p>Output is redirected into the file <i>output_file</i>. If absent, output goes to standard out.</p>
  </blockquote>
<p>-k <i> edit_distance</i></p>
  <blockquote>
<p>The maximum number of insertions, deletions, and substitutions permitted in any primer alignment. If absent, edit distance 0 is assumed.&nbsp;</p>
  </blockquote>
<p>-K <i>mismatches</i></p>
  <blockquote>
<p>The maximum number of mismatches permitted in any primer alignment. </p>
  </blockquote>
<p>-r&nbsp;</p>
  <blockquote>
<p> Search for the reverse complements of the primers too.</p>
  </blockquote>
<p>-<i>x l&nbsp;</i></p>
  <blockquote>
    <p>Length of exact seed or word size, ala blast, required by any primer 
    alignment. Can be combined with other options.</p>
  </blockquote>
<p>-s (<i> l | ~l </i>)<i>&nbsp;</i></p>
  <blockquote>
    <p> Constrain the first<i> l </i>primer characters to match exactly; any insertions, deletions or substitutions must occur after position<i>
    l</i>. The reverse complement of a primer must also have its first <i> l </i> characters
    match exactly. Note that a wildcard match is considered an exact match. With
    the ~ modifier, the first <i>l</i> primer characters are constrained to
    match inexactly, the remaining characters must match exactly.&nbsp;</p>
  </blockquote>
<p>-e (<i> l | ~l </i>)<i>&nbsp;</i></p>
  <blockquote>
    <p> Constrain the last <i> l </i>primer characters to match exactly; any insertions, deletions or substitutions must occur before position<i>
    l</i>. The reverse complement of a primer must also have its last<i> l</i> characters
    match exactly. Note that a wildcard match is considered an exact
    match.&nbsp; With the ~ modifier, the last <i>l</i> primer characters are
    constrained to match inexactly, the remaining characters must match exactly.</p>
  </blockquote>
<p>-5<i>  </i>(<i> l | ~l </i>)</p>
  <blockquote>
<p> Constrain the<i> l </i>primer characters at the 5' end of the primer to match
exactly; any insertions, deletions or substitutions must occur after position<i> l
</i> from the 5' end of the primer. The reverse complement of a primer must also
have the<i> l </i>characters at its 5' end match exactly. Note that a wildcard
match is considered an exact match. With the ~ modifier, the <i>l</i> primer
characters at the 5' end of the primer are constrained to match inexactly, the
remaining characters must match exactly.&nbsp;</p>
  </blockquote>
<p>-3<i>  </i>(<i> l | ~l </i>)&nbsp;</p>
  <blockquote>
<p> Constrain the<i> l </i>primer characters at the 3' end of a primer to match
exactly; any insertions, deletions or substitutions must occur after position<i> l
</i> from the 3' end of the primer. The reverse complement of a primer must also have the<i> l
</i> characters at its 3' end match exactly. Note that a wildcard match is
considered an exact match. With the ~ modifier, the <i>l</i> primer characters
at the 3' end of the primer are constrained to match inexactly, the remaining
characters must match exactly.&nbsp;</p>
  </blockquote>
<p>-w&nbsp;</p>
  <blockquote>
<p> Respect IUPAC ambiguity codes as wildcards, in both the sequence database <i>and</i>
the primers. A symbol from the sequence database is considered a wildcard match
to a primer symbol if either set of represented DNA symbols contains the other.
The only exception is that a N in the sequence database does not match any
primer symbol. Note: this is almost certainly what you want, as long stretches
of Ns are often used to indicate gaps in assembled sequence.&nbsp;</p>
  </blockquote>
<p>-W&nbsp;</p>
  <blockquote>
<p>Respect IUPAC ambiguity codes as wildcards, in both the sequence database <i>and</i>
the primers. A symbol from the sequence database is considered a wildcard match
to a primer symbol if either set of represented DNA symbols contains the other.
Also respects Ns in the sequence databases.&nbsp;&nbsp;</p>
  </blockquote>
<p>-u&nbsp;</p>
  <blockquote>
<p>Force all primers to uppercase characters.</p>
  </blockquote>
<p>-M <i> max</i></p>
  <blockquote>
<p>Stop counting primer occurrences once a primer has been seen <i> max</i>
times.</p>
  </blockquote>
<p>-A <i> format</i></p>
  <blockquote>
<p>Output format for primer alignments. See <b> Output Format</b> below. If present, alignments will be output.</p>
  </blockquote>
<p>-C <i> format</i></p>
  <blockquote>
<p>Output format for primer counts. See <b> Output Format</b> below. If present,
counts will be output.</p>
  </blockquote>
<p>-R <i> report_interval</i></p>
  <blockquote>
<p>Usually, <b> primer_match</b> accumulates many matches before taking the time
to output alignments. This reduces the running time tremendously. However, if you are debugging or want reassurance that
<b> primer_match</b> is actually doing something, setting <i> report_interval</i>
to 1 will force <b>primer_match</b> to report alignments as they are found.</p>
  </blockquote>
<p>-E <i> eos</i></p>
  <blockquote>
<p>Consider the sequence character with ascii code <i> eos</i> to represent the
end of the sequence in a FASTA entry. This character can never be part of an alignment, except if explicitly included in a primer sequence.
By default, 12 (new line) is considered the end of sequence character. The end of sequence character is inserted by
<b> compress_seq</b>.</p>
  </blockquote>
<p>-D ( <i>0</i> | <i>1</i> | <i>2</i> | <i>3</i> | <i>4</i> )&nbsp;</p>
  <blockquote>
<p>Select the sequence database pre-processing strategy. The default, 0, will
choose the fastest strategy, based on the pre-processing done, or not done, by <b>compress_seq</b>.</p>
<ol>
  <li>Sequence database has not been pre-processed.</li>
  <li>Sequence database has been indexed by <b>compress_seq</b>. This is the
    default behavior of <b>compress_seq</b>.</li>
  <li>Sequence database has been indexed and normalized by <b>compress_seq</b>,
    using the option <u>-n true</u>.</li>
  <li>Sequence database has been indexed, normalized and compressed by <b>compress_seq</b>,
    using the option <u>-z true</u>.</li>
</ol>
<p>Given the availability of pre-processed sequence database files, option 3 is
selected first, then option 4, then option 2, then option 1. This will typically
represent the fastest possible run time.&nbsp;</p>
  </blockquote>
<p>-B</p>
<blockquote>
  <p>Use buffered standard I/O rather than mmap to stream through the sequence
  database. On some platforms, where the use of mmap is somewhat unpredictable,
  this option may make it possible to run <b>primer_match</b> reliably.&nbsp;</p>
</blockquote>
<p>-v&nbsp;</p>
<blockquote>
  <p>Verbose (version &amp; diagnostic) output.</p>
</blockquote>
<p>-h&nbsp;</p>
  <blockquote>
<p> Command-line help.</p>
  </blockquote>
</blockquote>
<h3>Output Format</h3>
<blockquote>
<p>The default alignment output format is</p>
  <blockquote>
<pre>&gt;defline
sequence start end edits
alignment
primer index rc?</pre>
  </blockquote>
<p>where <u> defline</u> is the FASTA header line of the sequence entry containing the
alignment; <u> sequence</u> is the aligned sequence from the sequence database; <u>start</u> and
<u> end</u> are the space based start and end positions of the aligned sequence in the sequence entry;
<u> edits</u> is the number of insertions, deletions, and substitutions in the alignment;
<u> alignment</u> is a series of alignment characters indicating match, insertion, deletion or substitution at
each position of the alignment; <u> primer</u> is the aligned primer sequence; <u>index</u> is the index of this primer in the primer input set; and
<u> rc?</u> is &quot;REVCOMP&quot; if the primer matched in its reverse complement form.</p>
<p>For example</p>
  <blockquote>
<pre>&gt;CCO_UID:219000002141424:BAC_UID:human_12212001_reproc:LEN:33337
AGATCGCAGGTACATAAATGCTTCT 20115 20140 0
|||||||||||||||||||||||||
AGATCGCAGGTACATAAATGCTTCT 3242
&gt;CCO_UID:219000002142926:BAC_UID:human_12212001_reproc:LEN:2262
CCCATTCAGTCTTTCTTTTAAAAACATTTATTTTTAATTCAT 1671 1713 0
||||||||||||||||||||||||||||||||||||||||||
CCCATTCAGTCTTTCTTTTAAAAACATTTATTTTTAATTCAT 4781 REVCOMP</pre>
  </blockquote>
<p>and</p>
  <blockquote>
<pre>&gt;gi|683734|gb|U20581.1|MFU20581 Macaca fascicularis endothelin 3 mRNA
CAGCCAGATCTGAG 44 58 1
|||*||||||||||
CAGTCAGATCTGAG 3
&gt;gi|9967394|dbj|AB047965.1| Macaca fascicularis brain cDNA
CTCAGATCTGA-TG 1569 1582 1
|||||||||||v||
CTCAGATCTGACTG 3 REVCOMP</pre>
  </blockquote>
<p>and</p>
  <blockquote>
<pre>&gt;gi|21320903|dbj|AB059653.1| Macaca fascicularis PGDH1 mRNA 
 TGGATAATTTTT 2338 2350 1
 +++|^|+||||+
 WRRA-AWTTTTW 13
&gt;gi|21320905|dbj|AB059654.1| Macaca fascicularis PGDH2 mRNA 
 ACCGAGGAGGA 502 513 1
 ||*||+|||||
 ACAGAKGAGGA 11
&gt;gi|21320905|dbj|AB059654.1| Macaca fascicularis PGDH2 mRNA 
 AGCTG-GTGGG 512 522 1
 |||||v|||||
 AGCTGYGTGGG 18
&gt;gi|7593035|dbj|AB041420.1| Gorilla gorilla gene for alpha-1
 CGCCRGCACGAGTT 596 610 1
 ||||+|||^|||||
 CGCCAGCA-GAGTT 2
</pre>
  </blockquote>
<p>The default counts output format is</p>
  <blockquote>
<pre>index rc? primer count ( 0-count 1-count ... )</pre>
  </blockquote>
  <p>where <u> index</u> is the index of the primer; <u> rc?</u> is &quot;R&quot; for the reverse
complement of the primer and &quot;F&quot; otherwise; <u> primer</u> is the sequence of
the primer if <u> rc?</u> is &quot;F&quot; and the sequence of the primer's reverse
complement if <u>rc?</u> is &quot;R&quot;; <u> count</u> is the number of occurrences of primer in the
sequence database; and <u><i>k</i>-count</u> is the number of occurrences of primer in the
  sequence database with <i> k</i> insertions, deletions, or substitutions.</p>
<p>For example</p>
  <blockquote>
<pre>1 F TTACGGGCAGCTCA 9 ( 6 3 )
1 R TGAGCTGCCCGTAA 0 ( 0 0 )
2 F CCTTGCCAGTCAGATC 23 ( 8 15 )
2 R GATCTGACTGGCAAGG 0 ( 0 0 )
3 F CAGTCAGATCTGAG 15 ( 2 13 )
3 R CTCAGATCTGACTG 6 ( 0 6 )</pre>
  </blockquote>
<p>The command line options <u> -A</u> and <u> -C</u> give the user explicit control over the
output of alignments and counts respectively. Each format string contains conversion characters, which specify pieces of the alignment or count
output.</p>
<p>Alignment format conversion characters:</p>
</blockquote>
<div align="center">
  <center>
  <table border="0" width="100%">
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%h&nbsp;</td>
      <td width="89%" valign="top" align="left">FASTA header (defline) of the sequence entry containing the alignment.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%H</td>
      <td width="89%" valign="top" align="left"> First &quot;word&quot; of the FASTA
        header (defline) of the sequence entry containing the alignment. The first word is everything up to (but not
        including) the first whitespace character of the defline.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%f</td>
      <td width="89%" valign="top" align="left"> Index of the FASTA entry containing
        the alignment.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%s</td>
      <td width="89%" valign="top" align="left"> Start position of the alignment
        within the FASTA entry (space based).</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%e</td>
      <td width="89%" valign="top" align="left"> End position of the alignment in the
        FASTA entry (space based).</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%l</td>
      <td width="89%" valign="top" align="left"> Length of the alignment.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%5</td>
      <td width="89%" valign="top" align="left"> Position of the 5' end of the alignment in the sequence entry (space
        based).</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%3</td>
      <td width="89%" valign="top" align="left"> Position of the 3' end of the alignment in the sequence entry (space
        based).</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%S</td>
      <td width="89%" valign="top" align="left"> Start position (absolute) of the alignment in the sequence database.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%E&nbsp;</td>
      <td width="89%" valign="top" align="left"> End position (absolute) of the alignment in the sequence database.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%i</td>
      <td width="89%" valign="top" align="left"> Index of the aligned primer.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%d</td>
      <td width="89%" valign="top" align="left">Edit distance (number of insertions, deletions, substitutions) of the
        alignment.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%p</td>
      <td width="89%" valign="top" align="left"> The (forward) sequence of the primer, whether it was found in its
        forward or reverse complement form.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%P</td>
      <td width="89%" valign="top" align="left"> The FASTA header (defline) of
        the primer, if the primers came from a FASTA format file. Otherwise,
        &quot;&quot;.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%I&nbsp;</td>
      <td width="89%" valign="top" align="left">STS id (first column) of primer 
      entry, if the primers came from a UniSTS format file. Otherwise, &quot;&quot;.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%L&nbsp;</td>
      <td width="89%" valign="top" align="left">STS length for primer entry, if 
      the primers came from a UniSTS format file. Otherwise, &quot;&quot;.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%a&nbsp;</td>
      <td width="89%" valign="top" align="left">STS accession for primer entry, 
      if the primers came from a IUniSTS format file. Otherwise, &quot;&quot;.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%O&nbsp;</td>
      <td width="89%" valign="top" align="left">STS organism for primer entry, 
      if the primers came from a UniSTS format file. Otherwise, &quot;&quot;.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%&amp;&nbsp;</td>
      <td width="89%" valign="top" align="left">Alternative STS accessions for 
      primer entry, if the primers came from a UniSTS format file. Otherwise, 
      &quot;&quot;.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%X&nbsp;</td>
      <td width="89%" valign="top" align="left">STS chromosome for primer entry, 
      if the primers came from a UniSTS format file. Otherwise, &quot;&quot;.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%q</td>
      <td width="89%" valign="top" align="left"> The primer sequence of the alignment.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%Q</td>
      <td width="89%" valign="top" align="left">The primer sequence of the alignment, with alignment characters to
        indicate an insertion.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%t</td>
      <td width="89%" valign="top" align="left"> The aligned sequence from the sequence database.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%T</td>
      <td width="89%" valign="top" align="left"> The aligned sequence from the sequence database, with alignment
        characters to indicate deletion.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%A</td>
      <td width="89%" valign="top" align="left"> The string of alignment characters indicating exact match, insertion,
        deletion and subsitution at each position of the alignment.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%r</td>
      <td width="89%" valign="top" align="left"> &quot;F&quot; if the forward form of the primer was found, &quot;R&quot;
if the reverse complement form of the primer was found.</td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%R</td>
      <td width="89%" valign="top" align="left">
<p> &quot; REVCOMP&quot; if the reverse complement form of the primer was
found, &quot;&quot; otherwise.</p>
      </td>
    </tr>
    <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%%</td>
      <td width="89%" valign="top" align="left"> Percent (%).</td>
    </tr>
  </table>
  </center>
</div>
<blockquote>
<p>The default alignment format is &quot;&gt;%h\n %T %s %e %d\n %A\n %Q %i%R\n&quot;.</p>
<p>Count format conversion characters:</p>
</blockquote>
<table border="0" width="100%">
  <tr>
    <td width="12%" valign="top" align="left"></td>
    <td width="8%" valign="top" align="left">%i</td>
    <td width="89%" valign="top" align="left"> The primer index.</td>
  </tr>
  <tr>
    <td width="12%" valign="top" align="left"></td>
    <td width="8%" valign="top" align="left">%p</td>
    <td width="89%" valign="top" align="left"> The (forward form of the) sequence of the primer.</td>
  </tr>
  <tr>
      <td width="12%" valign="top" align="left"></td>
      <td width="8%" valign="top" align="left">%P</td>
      <td width="89%" valign="top" align="left"> The FASTA header (defline) of
        the primer, if the primers came from a FASTA format file. Otherwise,
        &quot;&quot;.</td>
  </tr>
  <tr>
    <td width="12%" valign="top" align="left"></td>
    <td width="8%" valign="top" align="left">%q</td>
    <td width="89%" valign="top" align="left"> The forward or reverse complement form of the primer.</td>
  </tr>
  <tr>
    <td width="12%" valign="top" align="left"></td>
    <td width="8%" valign="top" align="left">%r</td>
    <td width="89%" valign="top" align="left">&quot;F&quot; for the
      forward form of the primer, &quot;R&quot; for the reverse complement form of the primer.</td>
  </tr>
  <tr>
    <td width="12%" valign="top" align="left"></td>
    <td width="8%" valign="top" align="left">%R</td>
    <td width="89%" valign="top" align="left"> &quot; REVCOMP&quot; for the reverse complement form of the primer,
&quot;&quot; otherwise.</td>
  </tr>
  <tr>
    <td width="12%" valign="top" align="left"></td>
    <td width="8%" valign="top" align="left">%c</td>
    <td width="89%" valign="top" align="left"> Count for primer or reverse complement.</td>
  </tr>
  <tr>
    <td width="12%" valign="top" align="left"></td>
    <td width="8%" valign="top" align="left">%C</td>
    <td width="89%" valign="top" align="left"> Space separated list of counts for edit distance 0, 1,
      etc.</td>
  </tr>
  <tr>
    <td width="12%" valign="top" align="left"></td>
    <td width="8%" valign="top" align="left">%+</td>
    <td width="89%" valign="top" align="left"> Plus (+) if the count for this primer exceeded the maximum count
      threshold.</td>
  </tr>
  <tr>
    <td width="12%" valign="top" align="left"></td>
    <td width="8%" valign="top" align="left">%%&nbsp;</td>
    <td width="89%" valign="top" align="left"> Percent (%).</td>
  </tr>
</table>
<blockquote>
<p>The default count format is &quot;%i %q %c%+ ( %C )\n&quot;.</p>
</blockquote>
<h3>See Also</h3>
<blockquote>
<p><b><a href="pcr_match.htm">pcr_match</a>, <a href="compress_seq.htm">compress_seq</a></b></p>
</blockquote>
<h3>Author</h3>
<blockquote>
<p>Nathan Edwards</p>
</blockquote>
</html>